index.js

```diff
@@ -173,10 +182,139 @@ app.get("/continentFilter", function (req, res, next) {
+    var cartId = req.body.cartId;
+    var itemId = req.body.item._id;
+    var quantity = req.body.quantity;
+    var imageUrl = req.body.imageUrl;
+
+
+    if(cartId === undefined) {
+        var uniqueId = _.uniqueId('cartId-');
+        cartModel.sessions.push({cartId : uniqueId, cart : [], imageUrl: imageUrl});
+        res.send({cart: []});
+    } else {
+        cartModel.add(cartId, itemId, quantity, imageUrl);
+        res.send({cart: cartModel.get(cartId)});
+    }
+});
+
+
+app.put('/cart', function(req, res) {
+    var cartId = req.body.cartId;
+    var itemId = req.body.itemId;
+    var quantity = req.body.quantity;
+
+    if(quantity >= 0) {
+       cartModel.put(cartId, itemId, quantity);
+       sendCartWithSubTotal(cartId, res);
+    } else {
+       res.send('Quantity is a negative.');
+    }
+});
+
+app.delete('/cart', function(req, res) {
+    var cartId = req.query.cartId;
+    var itemId = req.query.itemId;
+
+    if(cartModel.delete(cartId, itemId)) {
+        sendCartWithSubTotal(cartId, res);
+    } else {
+        res.status(404).send({error: 'Item not found'});
+    };
+});
+
+function makeUrlFromUserCart(userCart) {
+    var url = BASEAPIURL + "/items?$filter=_id eq '" + userCart[0].itemId + "'";
+
+    for (var i = 1; i < userCart.length; i++) {
+        url += " or _id eq '" + userCart[i].itemId + "'";
+    }
+
+    return url;
+}
+
+function sendCartWithSubTotal(cartId, res) {
+
+    var userCart = cartModel.cartById(cartId);
+
+    var subTotal = 0;
+
+    if(userCart.length !== 0) {
+        var url = makeUrlFromUserCart(userCart);
+
+        var requests = [function (callback) {
+            request(url, function (err, response, body) {
+                // JSON body
+                if (err) {
+                    console.log(err);
+                    callback(true);
+                    return;
+                }
+
+                obj = JSON.parse(body);
+                callback(false, obj);
+            });
+        }];
+
+        var resultArray = [];
+
+        async.parallel(requests, function (err, results) {
+            if (err) {
+                console.log(err);
+                res.send(500, "Server Error");
+                return;
+            }
+
+            var query = results[0].value;
+
+            for (var i = 0; i < userCart.length; i++) {
+                var queriedItem = query.find(function(item) {
+                    return item._id === userCart[i].itemId;
+                });
+
+                resultArray.push({
+                    item: queriedItem,
+                    quantity: userCart[i].quantity,
+                    imageUrl: userCart[i].imageUrl
+                });
+                subTotal += userCart[i].quantity * queriedItem.price;
+            }
+
+            res.send({cart: resultArray, subTotal: subTotal});
+        });
+    } else {
+        res.send({cart: [], subTotal: 0});
+    }
+}
+
+app.get('/cart', function(req, res) {
```

Comment:
this should be better:
`var cartId = req.query.cartId;

+    if(cartId === undefined){
+        var cartId = _.uniqueId('cartId-');
+        cartModel.sessions.push({cart: [], cartId: cartId });
+    } else {
+        var userCart = cartModel.cartById(cartId);
+
+        if(userCart === undefined){
+            cartModel.sessions.push({cartId: cartId, cart: []});
+        }
+    }
sendCartWithSubTotal(cartId, res);
`